### DISCUSSION TITLE
File System API (v1/v2) Feedback

### DISCUSSION THREAD

**Discussion by EndangeredMassa** [vercel/community#530] [Announcements]
# Introduction

Our engineering team have been working hard on our next iteration of the [File System API](https://vercel.com/docs/file-system-api). Version 1 and version 2 (current) have both exposed some shortcomings. Version 3 is currently in development.

# What is the File System API?

You may not know too much about the File System API (FS API), and that's okay! In reality, we expect framework authors to be more aware of the FS API than most developers. However, in summary, the FS API is an interface between frameworks and Vercel deployments that is [serialized](https://phil.tech/2015/serializing-api-output/) both to and from the file system. You can learn a lot more about the FS API in its current form from our docs linked above.

If you are a framework author or a framework community member looking to integrate with Vercel, you may use the FS API to make deploys seamless for your framework.

All Vercel platform features are supported by the FS API.

# Version Summary

[Version 1](https://vercel.com/docs/file-system-api/v1) supported static files and Node.js-only serverless functions, with some configuration options. It did not completely support all of Vercel’s features. In general, the usability left something to be desired.

[Version 2](https://vercel.com/docs/file-system-api/v2) (current) was inspired by Next.js' build outputs. It expanded on version 1, supporting edge functions and more configuration options. However, despite updating our documentation to explain all of the available functionality, feedback surround version 2's usability has been largely negative.

# Feedback

In order to be "third time lucky", we want to hear from you! What issues did you experience with version 1 or version 2 of the FS API? What feedback do you have? How can we improve the usability of the FS API in version 3?

**Comment by magne4000** [comment]
I already posted those on your Support platform, but this can be informative to anyone. Those issues have been merged under `#00074967`

## [v2] dataRoute is stripped from its extensions and build fails

I'm using the File System API prerenderer-manifest.json to build my integration for a project.

The `dataRoute` property documentation states the following:
> The name of a on-demand Page (Function) without its extension, which is used for updating the JSON of the statically generated Page. Can be used for client-side transitions, but can also be left as an empty String if not needed.

But it seems that even if I put the extension, it gets removed. Should be fine then? Nope.

My `dataRoute` files are formatted as such `index.pageContext.json`, and if I put `dataRoute: 'index.pageContext'`, it gets stripped to just 'index.json' when it tries to find the file. Same thing if I put `dataRoute: 'index.pageContext.json'`

## [v2] `dynamicRoutes` in prerender-manifest.json doesn't seem to work

See https://github.com/vercel/vercel/discussions/7607

**Reply by brillout** [reply]
Hi @EndangeredMassa, thanks for the outreach. We appreciate it.

These two issues are a blocker for us to release [vite-plugin-vercel](https://github.com/magne4000/vite-plugin-vercel). Is there a prospect that these two issues are being considered? Let us know if there is anything unclear, we are happy to elaborate.

We are eager to add first-class Vercel support for Vite.

CC @rauchg since you were asking how you can help :-).

**Reply by EndangeredMassa** [reply]
@brillout thanks for the feedback! We just published a new version of this that we called the Build Output API (v3). We changed how routes work a bit. Can you give that a try? https://vercel.com/docs/build-output-api/v3

Feedback on v3 is welcome here: https://github.com/vercel/community/discussions/577

**Comment by magne4000** [comment]
Disclaimer: I'm working on [vite-plugin-vercel](https://github.com/magne4000/vite-plugin-vercel).

Some other remarks/suggestions:
- `routes-manifest.json`
  - When using `rewrites` rules or `dynamicRoutes`, sometimes the destination endpoint gets the destination URL as `response.url`, sometimes it gets the source url instead. It seems to differ if the destination endpoint is under `pages/api/*` or a `pages/*`
  - `dynamicRoutes` is confusing. It seems to be (too) close to `next.js` concepts.
    - Is this linked to the prerender manifest in any way? If not, I have the feeling that `rewrites` and `dynamicRoutes` could be merged
- `[function].nft.json`: Some tools/bundler plugins to generate those would be great. Actually, the tools used internally to bundle the endpoints would be really useful. They should be part of the documentation
- `prerender-manifest.json`
  -  I feel that `routes` and `dynamicRoutes` should be merged, or at least have the same properties. ISR pages should behave the same wether they are prerendered at build time or not
  - I would **really** need support for file formats other than `.html` or `.json`. I could have a framework rendering client code (the one behind `dataRoute`) in `.xml` files. Or I want to generate an RSS/Atom feed, that's `.xml` instead of the `.html` one, and I find it being a really good use-case for ISR

**Reply by EndangeredMassa** [reply]
@magne4000 thanks for the feedback! We just published a new version of this that we called the Build Output API (v3). We simplified routing at the interface level. Can you give that a try? https://vercel.com/docs/build-output-api/v3

Feedback on v3 is welcome here: https://github.com/vercel/community/discussions/577

**Comment by mashaal** [comment]
We've had an open issue on supporting Vercel in our Deno web-app framework for a few months now. I had some trivial success using `vercel-deno`, but ultimately had to pause because of the uncertainty of how to manage static files and how to run a build script with Deno (not Node).

https://github.com/vercel-community/deno/issues/95
https://github.com/vercel-community/deno/pull/94#issuecomment-1071362784

I think a workflow and documentation built around GitHub actions as a first-class citizen would be both more clear and more flexible for framework authors. For example, this 3rd party action _may_ do what we need, but due to the above, it's not currently high on our tracker to investigate. https://github.com/marketplace/actions/vercel-action#skip-vercels-build-step

Our rather stale issue: https://github.com/exhibitionist-digital/ultra/issues/9

**Reply by mashaal** [reply]
For reference, Deno Deploy offers a Deno-maintained GitHub Action. Static files are resolved automatically, and it's a (relatively) clear build pipeline

https://github.com/denoland/deployctl/blob/main/action/README.md
https://deno.com/deploy/docs/serve-static-assets

There are some limitations with Deno Deploy that I don't believe would be an issue on Vercel, so I would be keen to test v3 in whatever shapes it's in. @rauchg

**Reply by TooTallNate** [reply]
Hey @mashaal, are you on the Deno Discord? I'd love to chat with you more about this and get your framework natively supported on Vercel.

**Reply by mashaal** [reply]
@TooTallNate - Yes, mashaal#9189 - Thanks so much!

**Reply by EndangeredMassa** [reply]
@mashaal We just published a new version of this that we called the Build Output API (v3). Can you give that a try? https://vercel.com/docs/build-output-api/v3

Feedback on v3 is welcome here: https://github.com/vercel/community/discussions/577

**Comment by swamidass** [comment]
In principle the file based system is good, but in practice there were some real difficulties to using and debugging unexpected behavior.

In particular...

more visibility into the packaging process is needed. I had a serverless function that was not getting bundled correctly, but it wasn't clear what was being packaged or not. For comparison, Netlify provided enough visibility (lambda packages are built locally) into this in their system to quickly identify and correct problems.

Better documentation on the precise format of the configuration json would also be helpful, perhaps doing schema validation with informative error reporting. I recall a lot of uncertainty on how precisely to configure dependencies for packaging. Coupled with the inability to see the actual build packages (or directly configure the nft files), this made for a difficult problem to solve, at least for me!
